buildSrc/src/main/java/tech/pegasys/teku/repackage/Repackage.java

```diff
@@ -0,0 +1,200 @@
+import java.util.stream.Stream;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
+import java.util.zip.ZipOutputStream;
+import org.apache.commons.compress.archivers.ArchiveEntry;
+import org.apache.commons.compress.archivers.tar.TarArchiveEntry;
+import org.apache.commons.compress.archivers.tar.TarArchiveInputStream;
+import org.apache.commons.compress.archivers.tar.TarArchiveOutputStream;
+import org.apache.commons.compress.compressors.gzip.GzipCompressorInputStream;
+import org.apache.commons.compress.compressors.gzip.GzipCompressorOutputStream;
+
+public class Repackage {
+  /**
+   * To support reproducible builds, repackage the provided distribution file with modification
+   * times set to the provided date. This will replace the original distribution file.
+   *
+   * @param distFile The distribution file to repackage.
+   * @param date The files will have this modification date.
+   * @throws IOException
+   */
+  public static void repackage(String distFile, Date date) throws IOException {
+    // Create a temporary directory to use as a working directory.
+    // We will delete this when we are finished.
+    Path tempDir = Path.of(Files.createTempDirectory("repackage").toFile().getAbsolutePath());
+
+    try {
+      Path distPath = Path.of(distFile);
+      Path tempDistPath = tempDir.resolve(distPath.getFileName());
+      FileTime fileTime = FileTime.fromMillis(date.getTime());
+
+      if (distFile.endsWith(".zip")) {
+        repackageZip(distPath, fileTime, tempDir);
+      } else if (distFile.endsWith(".tar.gz")) {
+        repackageTarGz(distPath, fileTime, tempDir);
+      } else {
+        throw new IllegalArgumentException("bad distribution");
+      }
+
+      // Set the modification date of the final distribution file.
+      Files.setLastModifiedTime(tempDistPath, fileTime);
+      // Replace the original distribution file with the repackaged one.
+      Files.copy(tempDistPath, distPath, COPY_ATTRIBUTES, REPLACE_EXISTING);
+    } finally {
+      // Delete the temporary directory.
+      // Thank you, SubOptimal: https://stackoverflow.com/a/35989142
+      try (Stream<Path> walk = Files.walk(tempDir)) {
+        walk.sorted(Comparator.reverseOrder()).map(Path::toFile).forEach(File::delete);
+      }
+    }
+  }
+
+  private static void repackageZip(Path zipDist, FileTime fileTime, Path tempDir)
+      throws IOException {
+    try (ZipFile zf = new ZipFile(zipDist.toString())) {
+      Enumeration<? extends ZipEntry> zipEntries = zf.entries();
+      zipEntries
+          .asIterator()
+          .forEachRemaining(
+              entry -> {
+                try {
+                  Path path = tempDir.resolve(entry.getName());
+                  if (entry.isDirectory()) {
+                    Files.createDirectories(path);
+                  } else {
+                    Files.createDirectories(path.getParent());
+                    Files.copy(zf.getInputStream(entry), path);
+                  }
```

Comment:
I could probably write each file directly to the new archive rather than expanding to a file. 

I think the code below should work (and seems to preserve executable status at least on my Mac):

```
  private static void repackageZip(Path zipDist, FileTime fileTime, Path tempDir)
      throws IOException {
    Path newZipDist = tempDir.resolve(zipDist.getFileName());
    try (final ZipFile source = new ZipFile(zipDist.toFile());
        final ZipOutputStream output = new ZipOutputStream(Files.newOutputStream(newZipDist))) {
      final Enumeration<? extends ZipEntry> sourceEntries = source.entries();
      while (sourceEntries.hasMoreElements()) {
        final ZipEntry sourceEntry = sourceEntries.nextElement();
        final ZipEntry outputEntry = new ZipEntry(sourceEntry);
        outputEntry.setLastModifiedTime(fileTime);
        outputEntry.setLastAccessTime(fileTime);
        output.putNextEntry(outputEntry);
        IOUtils.copy(source.getInputStream(sourceEntry), output);
      }
    }
  }

  private static void repackageTarGz(Path tarDist, FileTime fileTime, Path tempDir)
      throws IOException {
    Path newTarDist = tempDir.resolve(tarDist.getFileName());
    try (final TarArchiveInputStream source =
            new TarArchiveInputStream(
                new GzipCompressorInputStream(Files.newInputStream(tarDist)));
        final TarArchiveOutputStream target =
            new TarArchiveOutputStream(
                new GzipCompressorOutputStream(Files.newOutputStream(newTarDist)))) {
      ArchiveEntry entry;
      while ((entry = source.getNextEntry()) != null) {
        final TarArchiveEntry tarEntry = (TarArchiveEntry) entry;
        tarEntry.setModTime(fileTime);
        target.putArchiveEntry(tarEntry);
        IOUtils.copy(source, target);
        target.closeArchiveEntry();
      }
    }
  }
```

Reply:
Yeah wow that looks/works way better. Thank you.

Reply:
I did verify that it works as expected too. In retrospect, this makes sense. I didn't realize it was possible to modify it in place like that. I should have taken a step back to think about it.

Reply:
I didn't know either - I got very much nerd-sniped on this one. :)